File: 3.DFS:BFS/BOJ/(14502번)연구소-삼성 SW 역량 테스트 기출/[jeunghun] solution.swift
import Foundation

let input = readLine()!.components(separatedBy: " ").map { Int($0)! }
let n = input[0], m = input[1]
var lab = [[Int]]()
var empities = [(Int, Int)]()

for i in 0..<n {
    let data = readLine()!.components(separatedBy: " ").map { Int($0)! }
    for j in 0..<m {
        if data[j] == 0 {
            empities.append((i, j))
        }
    }
    
    lab.append(data)
}

var candidates = [[(Int, Int)]]()

for i in 0..<empities.count-2 {
    for j in i+1..<empities.count-1 {
        for k in j+1..<empities.count {
            candidates.append([empities[i], empities[j], empities[k]])
        }
    }
}

let dx = [-1, 1, 0, 0]    //상, 하, 좌, 우
let dy = [0, 0, -1, 1]    //상, 하, 좌, 우
var result = 0

for candidate in candidates {
    var board = lab
    for (x, y) in candidate {
        board[x][y] = 1    //벽 세우기
    }
    
    for i in 0..<n {
        for j in 0..<m {
            if board[i][j] == 2 {
                //BFS
                var q = [(i, j)]
                while !q.isEmpty {
                    let (x, y) = q.removeFirst()
                    
                    for t in 0..<4 {
                        let nx = x + dx[t]
                        let ny = y + dy[t]
                        
                        if nx < 0 || ny < 0 || nx >= n || ny >= m { continue }
                        
                        if board[nx][ny] == 0 {
                            q.append((nx, ny))
                            board[nx][ny] = 2
                        }
                    }
                }
            }
        }
    }
    
    var tmp = 0
    for i in 0..<n {
        tmp += board[i].filter { $0 == 0 }.count
    }
    
    if result < tmp {
        result = tmp
    }
}

print(result)

File: 3.DFS:BFS/BOJ/(14502번)연구소-삼성 SW 역량 테스트 기출/[jeunghun] solution.py
from itertools import combinations
from collections import deque
import copy

n, m = map(int, input().split())
lab = []    # 연구소
empties = []    # 빈 공간
result = 0
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

for i in range(n):
    data = list(map(int, input().split()))
    lab.append(data)
    for j in range(m):
        if data[j] == 0:
            empties.append((i, j))

candidates = combinations(empties, 3)

for candidate in candidates:
    board = copy.deepcopy(lab)
    for i in range(len(candidate)):
        board[candidate[i][0]][candidate[i][1]] = 1

    for i in range(n):
        for j in range(m):
            if board[i][j] == 2:
                q = deque([(i, j)])
                while q:
                    y, x = q.popleft()

                    for k in range(4):
                        nx = x + dx[k]
                        ny = y + dy[k]
                        if nx < 0 or ny < 0 or nx >= m or ny >= n:
                            continue

                        if board[ny][nx] == 0:
                            board[ny][nx] = 2
                            q.append((ny, nx))

    tmp = 0
    for i in range(n):
        tmp += board[i].count(0)

    if tmp > result:
        result = tmp

print(result)